[zichen] iP

zichen-3974 opened the pull request:
### DukePro

> “I put the pro in procrastinating. Just kidding go get your stuff done >:/” - DukePro

DukePro tracks your events, deadlines and todos, then yells at you to get it done. It's,

- test-based
- easy to use
- QUICK~~LY GET YOUR THINGS DONE~~ to use

All you 🫵 have to do is,

1. download it from [here](https://github.com/zichen-3974/ip/releases/tag/A-Jar).
2. double click it _or you could run it from your terminal with_ `java -jar duke.jar`
3. start adding your tasks!

And it's **FREE**! Now that's a great price. 💸💸

Features:

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [x] Filtering tasks by date or keywords

 --- 
If you're a Java programmer, you could use this to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Implemented greet, echo and exit for bot

Commit:
Added UserList class to store and print tasks

Commit:
Renamed UserList to Tasks, Task abstracted out as own class

Commit:
Added 3 types of tasks, Todos, deadlines, events.

Commit:
Added semi-automation testing for bot

Commit:
Added error handling such as invalid task number, empty input

zichen-3974 committed:
Tasks can now be deleted.

zichen-3974 committed:
Add storage ability for Duke

DukeData.txt file will be created if it does not exist and will store
all of users input data. Previously data will be lost after process
terminates. The text file now store previous input and is loaded in at
the start of each process.

zichen-3974 committed:
Fixed indentation

zichen-3974 committed:
Add date time recognition to Duke

Duke can now parse date and time instead of just saving it as a string.
User can also find deadlines due or events ongoing on a specified date.

zichen-3974 committed:
Merge branch 'branch-Level-7'

zichen-3974 committed:
Merge branch 'branch-Level-8'

zichen-3974 committed:
Fixed misaligned brackets

zichen-3974 committed:
Merge remote-tracking branch 'origin/add-gradle-support

zichen-3974 committed:
Set up Gradle

zichen-3974 committed:
Refactor code to extract out closely related classes

Storage of data, user interaction is all in Duke.java file.

Extract user interaction, user input parsing and data storage into
seperate classes. Add abstract class Command for different types
of commands to inherit from.

This reduces duplicated code and improves readability.

zichen-3974 committed:
Add JUnit tests for Deadline, MyDate, Todo

Manual testing for classes.

This process is tedious. Adding JUnit tests makes it more efficient.

More tests for other classes to be added in future.

zichen-3974 committed:
Add Javadoc for classes and methods

zichen-3974 committed:
Fixed variable names to match coding standard

zichen-3974 committed:
Add search function where user can filter tasks by keyword

User could only use list to view all tasks or filter tasks by date.

Given a large number of tasks, it is hard for the user to find tasks
they need to find.

A search function that takes in a keyword filters out tasks containing
the keyword.

zichen-3974 committed:
Merge branch 'branch-A-JavaDoc'

zichen-3974 committed:
Merge branch 'branch-Level-9'

zichen-3974 committed:
Merge branch 'branch-A-CodingStandard'

pzaiming commented on src/main/java/duke/tasks/TodoTest.java:
```diff
@@ -0,0 +1,24 @@
+        assertEquals(todo2.formatForFile(), "T|X|another description");
+
+    }
+}
```
May want to consider moving this Test file to another directiory ./test so that all the test files will be in the same place.

pzaiming commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,140 @@
+     * @return the corresponding command.
+     */
+    public static Command parse(String fullCommand) {
+        String commandType = fullCommand.split(" ")[0].toLowerCase();
```
Can consider using enum instead.

zichen-3974 replied:
Indeed! Thanks for the suggestion :)

pzaiming commented on src/main/java/duke/tasks/Tasks.java:
```diff
@@ -0,0 +1,150 @@
+    private List<Task> taskList;
+
+    public Tasks() {
+        this.taskList = new ArrayList<Task>(100);
```
Need not necessary specify the size of the ArrayList?

pzaiming commented on src/main/java/duke/commands/ChangeStatusCommand.java:
```diff
@@ -0,0 +1,35 @@
+            tasks.markTaskDone(taskNumber, false);
+        }
+    }
+}
```
Great use of OOP here to change the status of the task!

pzaiming reviewed (commented):
Great job! Code was easy to read and follow.

zichen-3974 committed:
Add GUI for Duke.

Duke did not have a GUI and is only accessible via the terminal.

Adding a GUI makes it more user friendly.

Let's update commands and duke to accomodate for a GUI.

zichen-3974 committed:
Merge branch 'branch-Level-10'

zichen-3974 committed:
Add java assertions to Duke getResponse() and IncorrectCommand execute()

zichen-3974 committed:
Improve code quality

Some methods have deep nesting and indented happy path.

Restructuring code will make it more readable.

Let's extract parts in deep nested methods and place them in a seperate
method and unindent happy paths as much as possible.

zichen-3974 committed:
Merge pull request #1 from zichen-3974/branch-A-Assertions

Add java assertions

zichen-3974 committed:
Merge branch 'master' into branch-A-CodeQuality

zichen-3974 committed:
Merge pull request #2 from zichen-3974/branch-A-CodeQuality

Improve code quality

zichen-3974 committed:
Create gradle.yml

Add Continuous Integration

zichen-3974 committed:
Add task duplication checker

Duplicate tasks can be accepted in the task list.

User may accidentally create too many copies of the same task. It may be
messy to look at.

Let's not allow duplicate tasks of the same type and give the user a
reminder when a duplicate task is entered.

zichen-3974 committed:
Update gradle.yml

zichen-3974 committed:
Delete .github/workflows directory

zichen-3974 committed:
Add Ui image

zichen-3974 committed:
Update README.md

zichen-3974 committed:
Fix wrong application name in build.gradle
